[stdlib] Mark `contains` methods on `Range/ClosedRange` transparent

lorentey opened the pull request:
These are really popular shorthands for implementing bounds checks, but currently they aren’t even marked `@inline(__always)`, so in larger functions they tend to remain outlined, defeating optimizations that would otherwise happen.

The corresponding variants on the partial range types are `@_transparent`; that helps unoptimized performance, so let’s apply the same attribute here.

rdar://151177326

lorentey committed:
[stdlib] Mark contains methods on Range/ClosedRange transparent

These are really popular shorthands for implementing bounds checks, but currently they aren’t even marked `@inline(__always)`, so in larger functions they tend to remain outlined, defeating optimizations that would otherwise happen.

The corresponding variants on the partial range types are `@_transparent`; that helps unoptimized performance, so let’s apply the same attribute here.

rdar://151177326

lorentey commented:
@swift-ci test

lorentey commented:
I'm not expecting any benchmark impact, but.

@swift-ci benchmark

benrimmington commented:
Should you also update:

* `contains(_:)` in the `PartialRangeFrom` extension?

* `~=(_:_:)` in the `RangeExpression` extension?

glessard commented:
@swift-ci please benchmark

glessard commented:
### Performance (x86_64): -O

**Regression**            | **OLD** | **NEW** | **DELTA** | **RATIO**
:---                      | ---:   | ---:   | ---:   | ---:     
ArrayAppendGenericStructs | 1390.0 | 1765.0 | +27.0% | **0.79x (?)**

### Code size: -O


### Performance (x86_64): -Osize

**Regression**             | **OLD** | **NEW** | **DELTA** | **RATIO**
:---                       | ---:   | ---:    | ---:   | ---:     
StringDistance.utf16.ascii | 115.15 | 127.056 | +10.3% | **0.91x (?)**

### Code size: -Osize


### Performance (x86_64): -Onone

**Improvement**                    | **OLD** | **NEW** | **DELTA** | **RATIO**
:---                               | ---:    | ---:    | ---:   | ---:     
ClosedRangeContainsRange           | 15858.0 | 10739.0 | -32.3% | **1.48x**
RangeContainsClosedRange           | 12057.0 | 10318.0 | -14.4% | **1.17x (?)**
ArrayAppendReserved                | 2786.0  | 2530.0  | -9.2%  | **1.10x (?)**
Data.init.Sequence.809B.Count.RE.I | 19137.0 | 17850.0 | -6.7%  | **1.07x (?)**

### Code size: -swiftlibs

lorentey commented:
> Should you also update:
> * `contains(_:)` in the `PartialRangeFrom` extension?

This is not actually relevant to bounds checking, but sure, why not have them all transparent!

* `~=(_:_:)` in the `RangeExpression` extension?

Oh yes indeed!

lorentey committed:
[stdlib] On second thought, making `ClosedRange.contains(Range)` transparent is a bad idea

lorentey committed:
[stdlib] Make PartialRangeFrom.contains transparent

Kudos to @benrimmington for spotting this

lorentey committed:
[stdlib] Add a transparent specialization of ~= for `Range`

Making the generic ~= transparent does not help unoptimized code; adding a specialization does.

lorentey committed:
[test] Verify that [Closed]Range.contains and Range.~= get inlined even in unoptimized builds

lorentey committed:
[test] Add rudimentary test for transparent bounds checking helpers

lorentey commented:
@swift-ci smoke test

lorentey commented:
@swift-ci smoke test

lorentey commented:
Touching `~=` was a mistake. I will take a look, but this is not worth all this, so I may end up reverting that part.

```
13:50:27 
 ******************** TEST 'Swift(linux-x86_64) :: Constraints/rdar105781521.swift' FAILED ********************
[...]
13:50:27  /home/build-user/swift/test/Constraints/rdar105781521.swift:13:17: error: unexpected note produced: overloads for '~=' exist with these partially matching parameter lists: (Substring, String)
13:50:27      case .first(true):
13:50:27                  ^
13:50:27  FAIL: Swift(linux-x86_64) :: Constraints/rdar105782480.swift (1298 of 19342)
13:50:27  /home/build-user/swift/test/Constraints/rdar105782480.swift:14:10: error: expected error not produced
13:50:27        // expected-error@-1 {{value of type 'MyEnum' has no member 'invalid'}}
13:50:27  ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
13:50:27  /home/build-user/swift/test/Constraints/rdar105782480.swift:15:30: error: incorrect message found
13:50:27        // expected-error@-2 {{'let' binding pattern cannot appear in an expression}}
13:50:27                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
13:50:27                               type 'Range<MyEnum>' has no member 'second'
13:50:27  
13:50:27  --
13:50:27  
13:50:27  FAIL: Swift(linux-x86_64) :: Constraints/patterns.swift (1332 of 19342)
13:50:27  ******************** TEST 'Swift(linux-x86_64) :: Constraints/patterns.swift' FAILED ********************
13:50:27  /home/build-user/swift/test/Constraints/patterns.swift:125:7: error: unexpected error produced: reference to member 'HairForceOne' cannot be resolved without a contextual type
13:50:27  case .HairForceOne: // expected-error{{type 'any HairType' has no member 'HairForceOne'}}
13:50:27        ^
13:50:27  /home/build-user/swift/test/Constraints/patterns.swift:125:40: error: incorrect message found
13:50:27  case .HairForceOne: // expected-error{{type 'any HairType' has no member 'HairForceOne'}}
13:50:27                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
13:50:27                                         binary operator '~=' cannot be applied to two 'any HairType' operands
13:50:27  FAIL: Swift(linux-x86_64) :: Constraints/result_builder_invalid_stmts.swift (1381 of 19342)
13:50:27  ******************** TEST 'Swift(linux-x86_64) :: Constraints/result_builder_invalid_stmts.swift' FAILED ********************
13:50:27  /home/build-user/swift/test/Constraints/result_builder_invalid_stmts.swift:17:10: error: unexpected note produced: overloads for '~=' exist with these partially matching parameter lists: (Substring, String)
13:50:27      case 1: // expected-error {{expression pattern of type 'Int' cannot match values of type 'String'}}
13:50:27           ^
13:50:27  /home/build-user/swift/test/Constraints/result_builder_invalid_stmts.swift:30:10: error: unexpected note produced: overloads for '~=' exist with these partially matching parameter lists: (Substring, String)
13:50:27      case 1: // expected-error {{expression pattern of type 'Int' cannot match values of type 'String'}}
13:50:27           ^
13:50:27  /home/build-user/swift/test/Constraints/result_builder_invalid_stmts.swift:44:10: error: unexpected note produced: overloads for '~=' exist with these partially matching parameter lists: (Substring, String)
13:50:27      case 1: // expected-error {{expression pattern of type 'Int' cannot match values of type 'String'}}
13:50:27           ^
13:50:27  /home/build-user/swift/test/Constraints/result_builder_invalid_stmts.swift:56:10: error: unexpected note produced: overloads for '~=' exist with these partially matching parameter lists: (Substring, String)
13:50:27      case 1: // expected-error {{expression pattern of type 'Int' cannot match values of type 'String'}}
13:50:27           ^
[etc]
```

lorentey committed:
[stdlib] Delete transparent `~=` overload; mark the regular one always-inlined instead

The new overload messes up diagnostics too much.

lorentey commented:
@swift-ci smoke test

lorentey commented:
@swift-ci smoke test

lorentey commented:
macOS runs look back in green this morning; Windows runs less so.

@swift-ci smoke test

lorentey commented:
🤡 

```
10:17:20  UNRESOLVED: lldb-api :: lang/swift/completion/TestSwiftREPLCompletion.py (2 of 411)
10:17:20 
 ******************** TEST 'lldb-api :: lang/swift/completion/TestSwiftREPLCompletion.py' FAILED ********************
```

lorentey commented:
@swift-ci smoke test

lorentey commented:
This again; it looks like it may actually be triggered by this PR, although it doesn't reproduce locally:

```
[2025-05-22T06:04:25.386Z] /Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/swift/test/SILOptimizer/performance-annotations.swift:38:11: error: unexpected error produced: this code performs reference counting operations which can cause locking
[2025-05-22T06:04:25.386Z]   return a[i].x
[2025-05-22T06:04:25.386Z]           ^
[2025-05-22T06:04:25.386Z] /Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/swift/test/SILOptimizer/performance-annotations.swift:43:10: error: unexpected note produced: called from here
[2025-05-22T06:04:25.386Z]   return noRTCallsForArrayGet(a, 0)
[2025-05-22T06:04:25.386Z]          ^
```

lorentey commented:
Right, `Array`'s subscript does not rely on `Range.contains` or `~=` to validate the index; as far as I can tell, this PR does not touch that code.

lorentey commented:
No, this doesn't locally reproduce on x86_64, either. I'll try rebasing next.

lorentey commented:
Meanwhile,

@swift-ci smoke test macOS platform

lorentey commented:
Nah, this still passes locally, even on current HEAD, on arm64 and x86_64 both.

lorentey commented:
```
[2025-05-23T02:17:56.403Z] FAIL: Swift(macosx-x86_64) :: SILOptimizer/performance-annotations.swift (8067 of 19086)
```

¯\\_(ツ)_/¯

lorentey commented:
@swift-ci test

lorentey committed:
[stdlib] ClosedRange.contains: Leave this force-inlined, not transparent

Making it transparent evidently induces new retain/release traffic in Array’s subscript, even though I can find no indication that `ClosedRange.contains` is ever called by that code path. Oh well.

lorentey commented:
I'm resorting to taking things out until array lookup regression is gone. First up, a1e51a1 replaces `@_transparent` with `@inline(__always)`, sacrificing debug performance.

lorentey commented:
```
[2025-06-05T23:52:18.674Z] PASS: Swift(macosx-x86_64) :: SILOptimizer/performance-annotations.swift (8110 of 19207)
```

Cool; so evidently somehow `Array.subscript` does invoke one of these APIs. I'm guessing it might be `~=`?

lorentey commented:
Right, macOS got a checkmark.

Of course, Linux would like some attention now:

```
[2025-06-06T00:24:46.474Z] ◇ Test run started.
[2025-06-06T00:24:46.474Z] ↳ Testing Library Version: 6.3-dev (ca81dffea6bd63d)
[2025-06-06T00:24:46.474Z] ↳ Target Platform: x86_64-unknown-linux-gnu
[2025-06-06T00:24:46.474Z] ◇ Suite "URL.Template Template" started.
[2025-06-06T00:24:46.474Z] ◇ Suite "AttributedString Index Tracking" started.
[2025-06-06T00:24:46.474Z] ◇ Suite "URL.Template Value" started.
[2025-06-06T00:24:46.474Z] ◇ Suite "URL.Template PercentEncoding" started.
[2025-06-06T00:24:46.474Z] ◇ Suite "URL.Template Expression" started.
[2025-06-06T00:24:46.474Z] ◇ Test stringRoundTrip(template:) started.
[2025-06-06T00:24:46.474Z] ◇ Test convertToNFCAndPercentEncode() started.
[2025-06-06T00:24:46.474Z] ◇ Test creating() started.
[2025-06-06T00:24:46.474Z] ◇ Test collapseRight() started.
[2025-06-06T00:24:46.474Z] ◇ Test insertionAtEmptyRange() started.
[2025-06-06T00:24:46.474Z] ◇ Test expressibleByLiteral() started.
[2025-06-06T00:24:46.474Z] ◇ Test normalizedAddingPercentEncoding_unreserved() started.
[2025-06-06T00:24:46.474Z] ◇ Test parsingWithMultipleNames() started.
[2025-06-06T02:28:53.851Z] Cancelling nested steps due to timeout
[2025-06-06T02:28:53.854Z] ◇ Test invalidInputRanges() startedSending interrupt signal to process
[2025-06-06T02:29:11.487Z] script returned exit code 143
```
